libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routeline/RouteLineComponent.kt

```diff
@@ -9,66 +9,42 @@ import com.mapbox.maps.plugin.gestures.gestures
 import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener
 import com.mapbox.maps.plugin.locationcomponent.location
 import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.toNavigationRoutes
 import com.mapbox.navigation.core.MapboxNavigation
-import com.mapbox.navigation.core.directions.session.RoutesObserver
-import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
 import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
-import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.dropin.component.routefetch.RoutesAction
+import com.mapbox.navigation.dropin.component.routefetch.RoutesViewModel
+import com.mapbox.navigation.dropin.extensions.flowRouteProgress
+import com.mapbox.navigation.dropin.extensions.flowRoutesUpdated
 import com.mapbox.navigation.ui.maps.route.line.MapboxRouteLineApiExtensions.findClosestRoute
-import com.mapbox.navigation.ui.maps.route.line.MapboxRouteLineApiExtensions.setRoutes
+import com.mapbox.navigation.ui.maps.route.line.MapboxRouteLineApiExtensions.setNavigationRouteLines
 import com.mapbox.navigation.ui.maps.route.line.api.MapboxRouteLineApi
 import com.mapbox.navigation.ui.maps.route.line.api.MapboxRouteLineView
 import com.mapbox.navigation.ui.maps.route.line.model.MapboxRouteLineOptions
-import com.mapbox.navigation.ui.maps.route.line.model.RouteLine
+import com.mapbox.navigation.ui.maps.route.line.model.NavigationRouteLine
 import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.ifNonNull
 import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.launch
 
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 internal class RouteLineComponent(
     private val mapView: MapView,
     private val options: MapboxRouteLineOptions,
+    private val routesViewModel: RoutesViewModel,
+    private val routeLineApi: MapboxRouteLineApi = MapboxRouteLineApi(options),
+    private val routeLineView: MapboxRouteLineView = MapboxRouteLineView(options)
 ) : MapboxNavigationObserver {
```

Comment:
Should extend from `UIComponent`

Reply:
Using a shared coroutine scope means that any UIComponent can cancel all of the components.  This will lead to unpredictable side effects.  The `RouteLineComponent` only needs to observe `MapboxNavigation` events therefore it is a `MapboxNavigationObserver`.

Reply:
We're trying to build consistency in DropInUI. If there is an issue with the coroutine scope in the `UIComponent`, let's create a different pull request and address the issue with `UIComponent`. These classes are new so it is expected to have some issues to fix

Agreeing with @abhishek1508 here, this should be a `UIComponent`

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/5562

@cafesilencio Will you propose a test that reproduces the scenario for concern?

Reply:
My mistake for some reason my head was seeing it a static so there was only one instance in the parent class that all the children were using.

Reply:
@kmadsen why is it a lateinit var?  It should be immutable via a "by lazy {}"

Reply:
>  why is it a lateinit var? It should be immutable via a "by lazy {}"

After it is detached, we want to be able to attach again. Here's a test that will fail if it is lazy initialized. (side note: the 2 happens multiple times because the TestUIComponent has a `StateFlow`. The UIComponent can decide how state is held)

``` kotlin
    @Test
    fun `should be able to re-attach`() = runBlockingTest {
        val component = TestUIComponent(1)

        val mapboxNavigation = mockk<MapboxNavigation>()
        component.onAttached(mapboxNavigation)
        component.stateFlow.emit(2)
        component.onDetached(mapboxNavigation)
        component.onAttached(mapboxNavigation)
        component.stateFlow.emit(3)
        component.onDetached(mapboxNavigation)

        assertEquals(listOf(1, 2, 2, 3), component.captured)
    }
```